x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -69,75 +56,150 @@ interface IWaterfallItemBase {
-  docType: 'span';
+  name: string;
+}
+
+interface IError extends ITraceItem {
+  error: APMError;
+  message?: string;
+  serviceColor?: string;
+}
+
+interface IAgentMark {
+  mark: string;
 }
 
-export type IWaterfallItem = IWaterfallItemSpan | IWaterfallItemTransaction;
+export type IWaterfallTransaction = IWaterfallItemBase<
+  ITransaction,
+  'transaction'
+>;
+export type IWaterfallSpan = IWaterfallItemBase<ISpan, 'span'>;
+export type IWaterfallError = IWaterfallItemBase<IError, 'error'>;
+export type IWaterfallAgentMark = IWaterfallItemBase<IAgentMark, 'agentMark'>;
+
+export type IWaterfallItem =
+  | IWaterfallTransaction
+  | IWaterfallSpan
+  | IWaterfallError
+  | IWaterfallAgentMark;
 
 function getTransactionItem(
   transaction: Transaction,
   errorsPerTransaction: TraceAPIResponse['errorsPerTransaction']
-): IWaterfallItemTransaction {
+): IWaterfallTransaction {
   return {
-    id: transaction.transaction.id,
-    parentId: transaction.parent && transaction.parent.id,
-    serviceName: transaction.service.name,
-    name: transaction.transaction.name,
+    docType: 'transaction',
+    doc: {
+      id: transaction.transaction.id,
+      transaction,
+      errorsCount: errorsPerTransaction[transaction.transaction.id] || 0,
+      parentId: transaction.parent?.id,
+      serviceName: transaction.service.name,
+      timestamp: transaction.timestamp.us,
+      name: transaction.transaction.name
+    },
     duration: transaction.transaction.duration.us,
-    timestamp: transaction.timestamp.us,
     offset: 0,
-    skew: 0,
-    docType: 'transaction',
-    transaction,
-    errorCount: errorsPerTransaction[transaction.transaction.id] || 0
+    skew: 0
   };
 }
 
-function getSpanItem(span: Span): IWaterfallItemSpan {
+function getSpanItem(span: Span): IWaterfallSpan {
   return {
-    id: span.span.id,
-    parentId: span.parent && span.parent.id,
-    serviceName: span.service.name,
-    name: span.span.name,
+    docType: 'span',
+    doc: {
+      id: span.span.id,
+      span,
+      parentId: span.parent?.id,
+      serviceName: span.service.name,
+      timestamp: span.timestamp.us,
+      name: span.span.name
+    },
     duration: span.span.duration.us,
-    timestamp: span.timestamp.us,
+    offset: 0,
+    skew: 0
+  };
+}
+
+function getErrorItem(error: APMError): IWaterfallError {
+  return {
+    docType: 'error',
+    doc: {
+      id: error.error.id,
+      error,
+      message: error.error.log?.message || error.error.exception?.[0]?.message,
+      parentId: error.parent?.id,
+      serviceName: error.service.name,
+      timestamp: error.timestamp.us
+    },
     offset: 0,
     skew: 0,
-    docType: 'span',
-    span
+    duration: 0
   };
 }
 
+function getAgentMarks(transaction?: Transaction): IWaterfallAgentMark[] {
+  const agent = transaction?.transaction.marks?.agent;
+  if (!agent) {
+    return [];
+  }
+  return Object.entries(agent).map(([name, ms]) => ({
+    docType: 'agentMark',
```

Comment:
I'm not sure if it was a good idea to include `agentMark`'s in waterfall helpers. it seems like it is only causing us to filter it out again and doing special handling. Not sure what the advantage is.

My thinking was, that it should act as a child of `transaction`, and inherit the skew of this. Similar to how a span inherits the skew of the parent.

Reply:
I can add the agent marks to the Transaction interface, like this. WDYT?
```js
export interface ITransaction extends ITraceItem {
  transaction: Transaction;	  transaction: Transaction;
  docType: 'transaction';	  errorsCount: number;
  errorCount: number;	  name: string;
  marks: AgentMark[]
}	
```